### ISSUE TITLE
Very slow and buggy?

### ISSUE THREAD

**Issue by hopeseekr** [OWASP/rbac#55]
Hi,

First, I want to say that I find the utter lack of online source code demos for PHPRBAC to be troubling. But now this github issue will be one of the primary example code snippets on the Internet, so I'm hoping you can get this to work by showing me the error of my ways!

The following code mostly works, except for two failures which you'll see below. However, it takes an incredibly long time (30-40 seconds on an Intel i7). I want to know 1) how to fix the failures, 2) how to increase the performance dramatically and 3) any other tips you can give me.

Thank you so much! Time is of the essence.

Output:

```
Success: User 1 has the 'intent to leave' permission.
Success: User 1 has the 'profile-blocks-1' permission.
Failure! User 1 has the 'profile-blocks-2' permission.
Success! 'survey items-3' is an unknown/invalid permission.
Success: User 2 has the 'intent to leave' permission.
Failure! User 2 has the 'profile-blocks-1' permission.
Success: User 2 has the 'profile-blocks-n' permission.
Success! 'survey items-3' is an unknown/invalid permission.
```

```
<?php
// PHP-RBAC Proof of Concept

require_once 'PhpRbac/autoload.php';
header('Content-Type: text/plain');

$rbac = new PhpRbac\Rbac();

$commonPerms = [
    'intent-to-leave',
    'referrals-public-view',
    'referrals-teams-view',
];

$role1Perms = [
    'access level-individual_aggregate'.
    'risk-indicator',
    'profile-blocks-1',
    'profile-blocks-4',
    'profile-blocks-n',
    'profile-items-2',
    'profile-items-n',
    'survey-blocks-2',
    'survey-blocks-4',
    'survey-blocks-n',
    'survey-items-2',
    'survey-items-n',
    'notes-public-view',
    'notes-teams-view',
    'contacts-public-view',
    'contacts-teams-view',
    'booking-public-view',
    'booking-teams-view',
];

$role2Perms = [
     'access-level_aggregate',
     'profile-blocks-2',
     'profile-blocks-4',
     'profile-blocks-n',
     'profile-items-2',
     'profile-items-n',
     'survey-blocks-2',
     'survey-blocks-4',
     'survey-blocks-n',
     'survey-items-2',
     'survey-items-n',
     'notes-public-create',
     'notes-teams-create',
     'contacts-public-view',
     'contacts-private-create',
     'contacts-teams-view',
     'booking-public-create',
     'booking-teams-create',
     'referrals-private-create',
];

$role3Perms = [
    'access-level_individual_aggregate',
    'risk-indicator',
    'referrals-public-create',
    'referrals-private-create',
    'referrals-teams-create',
    'receive-referrals',
];


$commonRoleId = $rbac->Roles->add('common_role', 'Common permissions for all users');

foreach ($commonPerms as $info) {
    $permId = $rbac->Permissions->add($info, 'Default description', $commonRoleId);
    $rbac->Roles->assign('common_role', $permId);
}

$role1Id = $rbac->Roles->add('role1', 'Role 1', $commonRoleId);

foreach ($role1Perms as $permName) {
    $rbac->Permissions->add($permName, 'Default description', $role1Id);
    $rbac->Roles->assign('role1', $permName);
}

$role2Id = $rbac->Roles->add('role2', 'Role 2', $commonRoleId);

foreach ($role2Perms as $permName) {
    $rbac->Permissions->add($permName, 'Default description', $role2Id);
    $rbac->Roles->assign('role2', $permName);
}


// Create User
$users = [
    [
        'id' => 1,
        'username' => 'admin',
        'roles' => [
            'role1'
        ]
    ],
    [
        'id' => 2,
        'username' => 'editor',
        'roles' => [
            'common_role',
            'role2'
        ]
    ],
];

// Assign the proper roles.
foreach ($users as $user) {
    if (!empty($user['roles']) && is_array($user['roles'])) {
        foreach ($user['roles'] as $role) {
            $rbac->Users->assign($role, $user['id']);
        }
    }
}


$tests = [
    $users[0]['id'] => [
        'intent to leave' => true,
        'profile-blocks-1' => true,
        'profile-blocks-2' => false,
        'survey items-3' => false,
    ],
    $users[1]['id'] => [
        'intent to leave' => true,
        'profile-blocks-1' => false,
        'profile-blocks-n' => true,
        'survey items-3' => false,
    ],
];


// Validate if a user has a particular permission.
foreach ($tests as $userId => $test) {
    foreach ($test as $permission => $hasPermission) {
        $permissionVerb = $hasPermission ? 'has' : 'doesn\'t have';
        $noPermissionVerb = $hasPermission ? 'doesn\'t have' : 'has';
        try {
            if ($rbac->check($permission, $userId) === $hasPermission) {
                echo "Success: User {$userId} $permissionVerb the '{$permission}' permission.\n";
            } else {
                echo "Failure! User {$userId} $noPermissionVerb the '{$permission}' permission.\n";
            }
        }
        catch(Exception $e) {
            $status = $hasPermission ? 'Failure' : 'Success';
            echo "$status! '{$permission}' is an unknown/invalid permission.\n";
        }
    }
}
```

**Comment by Kern046** [comment]
I tested your code, just after init my DB with install.php.

The first times I launched it, it took between 240ms ans 400ms with Nginx. But after, when my DB was bigger, it took more time, I imagine it is the reason why after many tries, your script is slow.

I have the following errors on my own : 

> Failure! 'intent to leave' is an unknown/invalid permission.
> Success: User 1 has the 'profile-blocks-1' permission.
> Failure! User 1 has the 'profile-blocks-2' permission.
> Success! 'survey items-3' is an unknown/invalid permission.
> Failure! 'intent to leave' is an unknown/invalid permission.
> Failure! User 2 has the 'profile-blocks-1' permission.
> Success: User 2 has the 'profile-blocks-n' permission.
> Success! 'survey items-3' is an unknown/invalid permission.

**Comment by hopeseekr** [comment]
1. My code doesn't use a database at all...
2. Anything CLOSE to 200 ms for a simple AUTH check is COMPLETELY unreasonable (!!).
3. I wrote my own RBAC solution (what I'm calling TinyRbac) that does auth checks in 0.005 ms. I don't know (and doubt) if it is as feature-complete as PHP-RBAC, but it suits my needs.

Just how fast do you think PHP-RBAC _should_ be with 2 groups and less than 60 permissions, anyway? The systems I'm designing will need dozens of groups and hundreds, if not thousands, of permissions.

**Comment by abiusx** [comment]
PHP-RBAC is designed for enterprise, meaning that it is built to support 10,000+ roles and permissions, and it shouldnt be slow.
200 ms doesn’t really show anything as it is not a benchmark and your database might need warmups in the first run.

But clear these for me:
1. how do you not use a database? PHP-RBAC works using SQL and databases!
2. Are you sure your solution is hierarchical rbac? the main point of PHP RBAC is that it supports hierarchies, though I see none in your code example.
3. Your example code is a little bloated. I can not figure out what you logically expect from it to be able to see if it works as intended or not, and why.

Please let us know more and we sure can help you!

**Comment by ghost** [comment]
i have tried the code
first point
- miss spelling "intent to leave" and "intent-to-leave"
- "survey items-3" is completely missing

why should i insert the hole list of permissions over and over again?

if i only check the permissions
the script takes 10ms or less "Intel Core-i3"

```
Success: User 1 has the 'intent-to-leave' permission.
Success: User 1 has the 'profile-blocks-1' permission.
Failure! User 1 has the 'profile-blocks-2' permission.
Failure! User 1 has the 'survey-items-2' permission.
Success: User 2 has the 'intent-to-leave' permission.
Failure! User 2 has the 'profile-blocks-1' permission.
Success: User 2 has the 'profile-blocks-n' permission.
Failure! User 2 has the 'survey-items-2' permission.
0.0045080184936523s
```

**Comment by abiusx** [comment]
I didn’t understand your question.

**Comment by ghost** [comment]
English is not my natural language.
I try something better to formulate.
I just wanted to say that the problems of hopeseekr by incorrect use or spelling errors came about.

It works great.
Even in our financial accounting the examination and processing of rights takes no longer than 3-4ms.

**Comment by cheeman** [comment]
well, I tried to use the library in my code, but unfortunately when permissions grow up to about 80000+ items, my server got an 502 error.

**Comment by cheeman** [comment]
Thanks. Firstly, I added at least 80000 permissions continuously using $rbac->Permissions->add() method, and then I try to test the performance. For example, I set up a school permission which follow, and put 80000 permissions into the school permission. Next, I tried to use $brac->Permissions->children(1) , expecting a result include "school" permission. But, 502 error occur. Then I retrieved the query sql of the Permission->children API, and "EXPLAIN" it in mysql. It shows that two query using all table scaning, which resulting in an performance problem.
